service/src/main/java/bio/terra/externalcreds/services/JwtUtils.java

```diff
@@ -0,0 +1,147 @@
+      var passportJwt = decodeJwt(passportJwtString);
+
+      var visaJwtStrings =
+          Objects.requireNonNullElse(
+              passportJwt.getClaimAsStringList(GA4GH_PASSPORT_V1_CLAIM),
+              Collections.<String>emptyList());
+
+      var visas =
+          visaJwtStrings.stream().map(s -> buildVisa(decodeJwt(s))).collect(Collectors.toList());
+
+      return new LinkedAccountWithPassportAndVisas.Builder()
+          .linkedAccount(linkedAccount)
+          .passport(buildPassport(passportJwt))
+          .visas(visas)
+          .build();
+    } else {
+      return new LinkedAccountWithPassportAndVisas.Builder().linkedAccount(linkedAccount).build();
+    }
+  }
+
+  private static GA4GHPassport buildPassport(Jwt passportJwt) {
+    var passportExpiresAt = getJwtExpires(passportJwt);
+
+    return new GA4GHPassport.Builder()
+        .jwt(passportJwt.getTokenValue())
+        .expires(passportExpiresAt)
+        .build();
+  }
+
+  private static GA4GHVisa buildVisa(Jwt visaJwt) {
+    JSONObject visaClaims = getJwtClaim(visaJwt, GA4GH_VISA_V1_CLAIM);
+    var visaType = visaClaims.get(VISA_TYPE_CLAIM);
+    if (visaType == null) {
+      throw new InvalidJwtException(String.format("visa missing claim [%s]", VISA_TYPE_CLAIM));
+    }
+    return new Builder()
+        .visaType(visaType.toString())
+        .jwt(visaJwt.getTokenValue())
+        .expires(getJwtExpires(visaJwt))
+        .issuer(visaJwt.getIssuer().toString())
+        .lastValidated(new Timestamp(Instant.now().toEpochMilli()))
+        .tokenType(determineTokenType(visaJwt))
+        .build();
+  }
+
+  private static Timestamp getJwtExpires(Jwt decodedPassportJwt) {
+    var expiresAt = decodedPassportJwt.getExpiresAt();
+    if (expiresAt == null) {
+      throw new InvalidJwtException("jwt missing expires (exp) claim");
+    }
+    return new Timestamp(expiresAt.toEpochMilli());
+  }
+
+  private static <T> T getJwtClaim(Jwt jwt, String claimName) {
+    T claim = jwt.getClaim(claimName);
+    if (claim == null) {
+      throw new InvalidJwtException(String.format("jwt missing claim [%s]", claimName));
+    }
+    return claim;
+  }
+
+  private static TokenTypeEnum determineTokenType(Jwt visaJwt) {
+    // https://github.com/ga4gh/data-security/blob/master/AAI/AAIConnectProfile.md#conformance-for-embedded-token-issuers
+    return visaJwt.getHeaders().containsKey(JKU_HEADER)
+        ? TokenTypeEnum.document_token
+        : TokenTypeEnum.access_token;
+  }
+
+  @VisibleForTesting
+  Jwt decodeJwt(String jwtString) {
+    try {
+      // first we need to get the issuer from the jwt, the issuer is needed to validate
+      var jwt = JWTParser.parse(jwtString);
+      var issuer = jwt.getJWTClaimsSet().getIssuer();
+      if (issuer == null) {
+        throw new InvalidJwtException("jwt missing issuer (iss) claim");
+      }
+      var jkuOption = Optional.ofNullable(((JWSHeader) jwt.getHeader()).getJWKURL());
+      if (jkuOption.isPresent()) {
```

Comment:
FWIW, you can simplify this a bit with:
```suggestion
      return jkuOption.map(jku ->
      if (jkuOption.isPresent()) {
        if (externalCredsConfig.getAllowedJwksUris().contains(jku)) {
          return ExternalCredsJwtDecoders.fromJku(jku).decode(jwtString);
        } else {
          throw new InvalidJwtException(
              String.format("URI [%s] specified by jku header not on allowed list", jku));
        }).orElseGet(() -> JwtDecoders.fromIssuerLocation(issuer).decode(jwtString))
```

Reply:
Thank you for the suggestion! For some reason, the map function doesn't like that an exception could be thrown inside of it and caught outside of it, so we might just leave this part as it currently is.

Reply:
fair enough!